Typed everything; removed redundant funcs

Xananax opened the pull request:
I had not noticed that there already was a Godot 4 branch, so I did my own. I still submit it here because I think it has several improvements that can be of interest:

- Typed everything (using inference wherever possible)
- Includes a Github action that creates a zipped archive with the plugin (without any redundant files), which makes for a cleaner asset repository submission
- Removed redundant `_get_tree_item_children` method (the regular `TreeItem.get_children()`does the same job in Godot 4)
- Corrected `_get_index_path()` (though it's not used anywhere, maybe we could remove it?)
- Corrected `is_inspection_enabled`
- Added icons support (see screenshot)

![image](https://user-images.githubusercontent.com/836082/213462499-f368af87-13b2-441c-898e-999dfaf944c8.png)

I also compared it to the current `godot4` branch to minimize changes

Zylann commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,21 @@
+      - name: Release
+        uses: softprops/action-gh-release@v1
+        with:
+          files: godot-editor-debugger.zip
```
Not sure if I want this. Usually I do assetlib releases using version tags, which is occasional, and this seems to make a new zip file and a new release on every push, which I don't want.

Xananax replied:
It only makes them on tagged pushes. So if you use 
```sh
git tag v2.1.0
```
then
```sh
git push && git push --tags
```
it will create a release; otherwise, it won't. It will also only create a release if the tag is in the form `v*.*.*`. Any other tag will be ignored.

The rule can be found towards the top of the file

```yaml
on:
  push:
    tags:
      - "v*.*.*"
````

Zylann replied:
Also I actually dont even have version branches in this plugin yet, as I'm not working on it often (Even in other repos I dont even use `git tag` but rather make a branch directly, because then I dont have to use any different git commands xD). I'm just not sure I want this for now

Xananax replied:
That's fair, I'll remove it

Zylann commented on addons/zylann.editor_debugger/dock.gd:
```diff
@@ -1,29 +1,29 @@
 extends Control
 
-const Util = preload("util.gd")
+const Util := preload("util.gd")
```
I don't think `:` is necessary in this case.

Xananax replied:
It's possible, I just keep it so it's easy to detect variables without a type by doing a project-wide search for `(var|const)\s+\w+\s*[^:]=`.

I'll remove it if it bothers you

Zylann commented on addons/zylann.editor_debugger/dock.gd:
```diff
@@ -1,29 +1,29 @@
-@onready var _label = get_node("VBoxContainer/Label")
-@onready var _tree_view : Tree = get_node("VBoxContainer/Tree")
-@onready var _save_branch_file_dialog = get_node("SaveBranchFileDialog")
+@onready var _popup_menu := get_node("PopupMenu") as PopupMenu
```
Usually when I type onready nodes, I put the type after `:` rather than using `as`, because from how I'm thinking, `as` means returning null without error if the object is not of the given type. Here I want it to fail if the type is incorrect (even though I havent tested if GDScript behaves that way yet)

Xananax replied:
That is correct, previously using the form `id:T = $Path`  did not give you autocomplete or static checks, but I just tested and in Godot 4 it seems it works. The `as` form is unecessary.

Zylann commented on addons/zylann.editor_debugger/dock.gd:
```diff
@@ -1,29 +1,29 @@
+@onready var _save_branch_file_dialog := get_node("SaveBranchFileDialog") as FileDialog
 
-const METADATA_NODE_NAME = 0
+const METADATA_NODE_NAME := 0
```
Typing constants should not be necessary (if it is, it's a bit annoying)

Xananax replied:
I'm really not sure, I do it out of consistency. Might not be useful.

Edit: just checked, consts seem to behave strictly like fully typed variables, so the `:` is not necessary.

Zylann commented on addons/zylann.editor_debugger/plugin.gd:
```diff
@@ -1,25 +1,44 @@
+# alternative class name as the second argument to avoid getting a "not found" icon
+# @param icon_name              For example, "Button", or "Node2D"
+# @param if_not_found_pick_this Generally, "Node". Leave blank to get a broken "not found" icon
+func get_theme_icon(icon_name: String, if_not_found_pick_this:="") -> Texture2D:
```
Why does this function need to be here?

Xananax replied:
It's required to use `get_editor_interface`,  which is the only reliable way I know to obtain editor icons for the nodes and such.

`get_editor_interface` is only available in addon scripts, so it needs to be there to be passed to the Dock, so the dock can create icons. It's not super clean, but I find it still simpler than other systems. Open to suggestions though.

Zylann replied:
Can't controls present in the editor tree access icons? I'd rather just having the thing there rather than being injected. The dock is a child of the editor interface so it should be able to access icons.

Xananax replied:
It should be possible yes. I'll test it.

Xananax replied:
Turns out, `get_theme_icon` is a method of `Control`. I'm unsure if it wasn't in 3.x and it changed, or if I've simply been doing this delegation dance for nothing all along.

Zylann commented on addons/zylann.editor_debugger/dock.gd:
```diff
@@ -67,24 +67,24 @@ func _update_tree():
-func _update_branch(root: Node, root_view: TreeItem):
-	if root_view.collapsed and root_view.get_first_child() != null:
+func _update_branch(root: Node, root_view: TreeItem) -> void:
+	if root_view.collapsed and root_view.get_child_count() != 0:
```
I looked it up, `get_child_count` is more expensive than `get_first_child`. It potentially builds a cache of all the children to then return its size, while `get_first_child` just returns one pointer in any case.

Xananax replied:
I see, will fix that

Zylann commented on addons/zylann.editor_debugger/plugin.gd:
```diff
@@ -1,25 +1,44 @@
-var _dock = null
+const Dock := preload("res://addons/zylann.editor_debugger/dock.gd")
+const DockScene: PackedScene = preload("res://addons/zylann.editor_debugger/dock.tscn")
+var _dock: Dock = null
```
Missing empty line

Zylann commented on addons/zylann.editor_debugger/plugin.gd:
```diff
@@ -1,25 +1,44 @@
 
-var _dock = null
+const Dock := preload("res://addons/zylann.editor_debugger/dock.gd")
+const DockScene: PackedScene = preload("res://addons/zylann.editor_debugger/dock.tscn")
```
Paths can be relative

Xananax commented:
Note: the "save branch as scene" button doesn't seem to work for me. I don't know if it's a regression or what, but the signal never fires in Godot 4 (I didn't test it in Godot 3).

I removed the button and replaced it with an item in the popup, which works. 

Should I try to fix this as part of this PR?

![image](https://user-images.githubusercontent.com/836082/213689341-4e8789f9-a5d2-46d3-8f19-42d519424f0c.png)


```gdscript
func _on_popup_menu_index_pressed(index: int) -> void:
	if index == 0:
		_on_SaveBranchAsSceneButton_pressed()


func _on_SaveBranchAsSceneButton_pressed() -> void:
	#_save_branch_as_scene_button.accept_event()
	_popup_menu.hide()
	_save_branch_file_dialog.popup_centered_ratio()
```
---

We're using this plugin at GDQuest to help us author editor plugins. To make it more useful for our case, we're adding the following right-click functions:

1. "Copy Path to clipboard", which copies the current path (like: `/root/@@16478/@@655/@@656//FileSystem/`)
2. "Copy Node Types to clipboard", which copies node class and name tuples (like `[["EditorNode", "@@16478"], ["Control", "@@655"], ["Panel", "@@656"], ["FileSystemDock", "FileSystem"]]`)

Are you interested in incorporating those, or should we maintain them in a fork?

And if yes, as part of this PR or a different one?

This is the entire code:

```gdscript
func _on_popup_menu_index_pressed(index: int) -> void:
	_popup_menu.hide()
	match index:
		0: # Save Branch as Scene
			_save_branch_file_dialog.popup_centered_ratio()
		1: # Copy Path to Clipboard
			var node_view := _tree_view.get_selected()
			var node := _get_node_from_view(node_view)
			DisplayServer.clipboard_set(node.get_path())
		2: # Copy node types to Clipboard
			var node_view := _tree_view.get_selected()
			var node := _get_node_from_view(node_view)
			var node_types := []
			while node.get_parent():
				var tuple := PackedStringArray([node.get_class(), node.name])
				node_types.append(tuple)
				node = node.get_parent()
			node_types.reverse()
			var node_types_str := "%s"%[node_types]
			DisplayServer.clipboard_set(node_types_str)
```

Zylann commented:
> I removed the button and replaced it with an item in the popup, which works.

Which popup? My memory is rusty.

> Should I try to fix this as part of this PR?

I'd prefer features to be done in separate PRs, yes.
Early comments on the code: don't use raw numbers for indices, prefer named constants. Also instead of `index_pressed`, use `id_pressed` (so visual order of items doesnt matter in the handler logic). A downside is that items may be setup in code, but I also prefer that to doing so in tscn.
I'm not sure what the goal of option 2 is. "Copy node types", but why? Also it doesnt only copy type, but a tuple of type and name, that feels a bit arbitrary.

Xananax commented:
> Which popup? My memory is rusty.

There's a PopupMenu that shows on right click, with a button "Save Current Branch as Scene". The button is visually cut off, and doesn't work when pressed. I'll pursue this in [a different PR](https://github.com/Zylann/godot_editor_debugger_plugin/pull/6).

Is there anything else that should be part of this PR?

Zylann commented:
Should be ok now, thanks!
